thrift/thrift0.13/src/test/java/com/linecorp/armeria/it/client/retry/RetryingRpcClientTest.java

```diff
@@ -326,34 +332,136 @@ void shouldGetExceptionWhenFactoryIsClosed() throws Exception {
+
+                                     if (param.ensureCancelBeforeFirstRequest) {
+                                         Thread.sleep(param.cancelDelayMillis);
+
+                                         assertThat(ctx.isCancelled()).isFalse();
+                                         assertThat(res.isDone()).isFalse();
+                                         assert param.cancelViaCtx;
+                                         ctx.cancel();
+                                         serviceRetryCountWhenCancelled.set(serviceRetryCount.get());
+
+                                         resFuture.complete(delegate.execute(ctx, req));
+                                     } else {
+                                         final RpcResponse nextRes = delegate.execute(ctx, req);
+                                         resFuture.complete(nextRes);
+
+                                         Thread.sleep(param.cancelDelayMillis);
+
+                                         assertThat(ctx.isCancelled()).isFalse();
+                                         assertThat(nextRes.isDone()).isFalse();
+
+                                         if (param.cancelViaCtx) {
+                                             ctx.cancel();
+                                         } else {
+                                             nextRes.cancel(true);
+                                         }
+
+                                         serviceRetryCountWhenCancelled.set(serviceRetryCount.get());
+                                     }
+
                                      return res;
                                  })
+                                 .rpcDecorator((delegate, ctx, req) -> {
+                                     context.set(ctx);
+                                     // Make sure we do not get cancelled while delaying the cancel.
+                                     ctx.setResponseTimeout(
+                                             TimeoutMode.EXTEND,
+                                             Duration.ofMillis(param.cancelDelayMillis + 5000)
+                                     );
+
+                                     return delegate.execute(ctx, req);
+                                 })
                                  .build(HelloService.Iface.class);
             when(serviceHandler.hello(anyString())).thenThrow(new IllegalArgumentException());
 
-            assertThatThrownBy(() -> client.hello("hello")).isInstanceOf(CancellationException.class);
+            assertThatThrownBy(() -> client.hello("hello"))
+                    .isOfAnyClassIn(CancellationException.class, ResponseCancellationException.class);
 
             await().untilAsserted(() -> {
-                verify(serviceHandler, only()).hello("hello");
+                assertThat(serviceRetryCountWhenCancelled.get()).isIn(serviceRetryCount.get(),
+                                                                      serviceRetryCount.get() - 1);
```

Comment:
It would be nice if we can make these assertions deterministic

e.g.
```suggestion
                if (param.ensureCancelBeforeFirstRequest || param.cancelDelayMillis > 0) {
                    assertThat(serviceRetryCountWhenCancelled.get()).isIn(serviceRetryCount.get());
                } else {
                    assertThat(serviceRetryCountWhenCancelled.get()).isIn(serviceRetryCount.get() - 1);
                }
```

Reply:
With `param.ensureCancelBeforeFirstRequest` we want to assert that zero attempts are made so I better put this in a separate if-branch.

You cannot rely on the delay whether there is another retry (`.isIn(serviceRetryCount.get() - 1)`) or not (`.isIn(serviceRetryCount.get() )`). If we do, this test gets brittle again.